Feature/message manager support max

mm-gmbd opened the pull request:
- Added “_readOptions” table that allows a user to define the step/skip
parameters for SPIFlashLogger.read method
- Added onReplayBegin/Complete handlers that allow a user to define
middleware for what happens when replaying data begins and is completed
- Added support for “retryOnInit” option that allowed a user to delay
calling “_scheduleRetryIfConnected” (in case there are several messages
that need to be sent on init — this can be memory intensive)
- Modified _onAck and _onFail so that it only catches messages that
have “loggerName” in the metadata — this assumes that this key being
present means that it was sent through RM (a better option may be to
just use the per-message onAck/onFail handlers)
- Include message metadata in what is saved to SPIFlash — don’t assume
that a user hasn’t included something important here

mm-gmbd committed:
Add replay middleware and ability to configure how to read from SPIFlashLogger

- Added “_readOptions” table that allows a user to define the step/skip
parameters for SPIFlashLogger.read method
- Added onReplayBegin/Complete handlers that allow a user to define
middleware for what happens when replaying data begins and is completed
- Added support for “retryOnInit” option that allowed a user to delay
calling “_scheduleRetryIfConnected” (in case there are several messages
that need to be sent on init — this can be memory intensive)
- Modified _onAck and _onFail so that it only catches messages that
have “loggerName” in the metadata — this assumes that this key being
present means that it was sent through RM (a better option may be to
just use the per-message onAck/onFail handlers)
- Include message metadata in what is saved to SPIFlash — don’t assume
that a user hasn’t included something important here

mm-gmbd commented:
Okay -- finally got this to open. @ppetrosh let me know what you think.

ppetrosh commented on ReplayMessenger.device.lib.nut:
```diff
@@ -55,6 +64,10 @@ class ReplayMessenger {
 
     constructor(options = {}) {
 
+        _onReplayBegin = {};
```
`_beforeReplay` and `_afterReplay` would probably be shorter while still clear

mm-gmbd replied:
Doesn't bother me either way :)

ppetrosh replied:
It's just that we already use before*/after* form in other libraries.

ppetrosh commented on ReplayMessenger.device.lib.nut:
```diff
@@ -93,6 +109,41 @@ class ReplayMessenger {
+        if (loggerName in _onReplayBegin) {
+          delete _onReplayBegin[loggerName]
+        }
+      } else if (callback == null) { //assume callback is provided for "default" logger
```
The block:

```
...
} else if (callback == null) { //assume callback is provided for "default" logger
    callback = loggerName;
    loggerName = "default";
} else ...
```

does nothing and makes no sense to me. It should either be removed or fixed. 

Or am I terribly missing anything?

mm-gmbd replied:
Oops - good catch. I think the comment describes what the purpose is, no? If the callback is "null" and the typeof the loggerName is *not* a string (i.e. it is a function), then the user is trying to provide a callback for the "default" logger. I'll update the code to make things clearer.

ppetrosh replied:
No it still doesn't make sense to me. I mean the code doesn't do what you expect it to do I guess. It doesn't set a handler. What it does is changing the local variables which are never used then.

ppetrosh commented on ReplayMessenger.device.lib.nut:
```diff
@@ -93,6 +109,41 @@ class ReplayMessenger {
+        if (loggerName in _onReplayComplete) {
+          delete _onReplayComplete[loggerName]
+        }
+      } else if (callback == null) { //assume callback is provided for "default" logger
```
The same as above (copy paste?).

I'd also suggest to get rid of this code duplication as we already got into trouble because of that.

mm-gmbd committed:
Update onReplayBegin/Complete

- rename onReplayBegin to beforeReplay (for brevity)
- rename onReplayComplete to afterReplay (for brevity)
- fix bug in before/afterReplay where we assume that the “default”
logger the target of the action (for both registering and
de-registering the callback)

mm-gmbd commented:
@ppetrosh - just updated. Review changes?

ppetrosh commented on ReplayMessenger.device.lib.nut:
```diff
@@ -93,6 +109,41 @@ class ReplayMessenger {
+    }
+
+    function setLoggerReadOptions(loggerName, options=null) {
+      if (typeof loggerName == "table" && options == null) {
```
This is a vicious practice to dynamically switch positions of the arguments! Let's please not do so!

a) it is very confusion and 
b) `setLoggerReadOptions(null, {...});` would make it fail.

If you want to have named optional parameters let's just use a map with two values. But I don't really think it's worth it.

ppetrosh commented on ReplayMessenger.device.lib.nut:
```diff
@@ -93,6 +109,51 @@ class ReplayMessenger {
+    }
+
+    function setLoggerReadOptions(loggerName, options=null) {
+      if (typeof loggerName == "table" && options == null) { //assume these are the read options provided for the "default" logger
```
This is a vicious practice to dynamically switch positions of the arguments! Let's please not do so!

a) it is very confusion and
b) setLoggerReadOptions(null, {...}); would make it fail.

If you want to have named optional parameters let's just use a map with two values. But I don't really think it's worth it.

ppetrosh replied:
Suggested a proposal above https://github.com/electricimp/ReplayMessenger/pull/7/files#diff-42394b1d52af90060c4c11545ae913daR140

ppetrosh replied:
```
function setLoggerReadOptions(options = null, loggerName = "default") {...}
```

which would allow us to get rid of this questionable block of code.

ppetrosh commented on ReplayMessenger.device.lib.nut:
```diff
@@ -93,6 +109,51 @@ class ReplayMessenger {
+        return;
+      }
+
+      if (typeof loggerName == "function" && callback == null) { //assume this is the callback provided for "default" logger
```
I don't mind making the logger name optional: `loggerName = "default"`, but I don't feel comfortable with shifting the parameter types. It's error prune and confusing for ourselves.

ppetrosh replied:
Another option would be to change the order of arguments and make the loggerName optional with default "default" value. What do you think?

ppetrosh replied:
So it should probably be:
```
function afterReplay(callback = null, loggerName = "default") {...}
```

mm-gmbd replied:
Hmm -- we could change the order of the arguments but that *feels* backwards if I'm setting a known logger. However, that seems to be the best path forward to get around the parameter switching considering the class *does* support the "default" logger.

ppetrosh commented on ReplayMessenger.device.lib.nut:
```diff
@@ -93,6 +109,51 @@ class ReplayMessenger {
         return _mm.send(messageName, data, null, _retryInterval, metadata);
     }
 
+    function beforeReplay(loggerName, callback = null) {
```
It's better to extract common code for `beforeReplay` and `afterReplay` into a separate function.

mm-gmbd committed:
Switch arguments to before/afterReplay

- Less error prone, no argument switching, simpler

mm-gmbd committed:
Modify arguments for setLoggerReadOptions

- Simplify setLoggerReadOptions by switching argument order

mm-gmbd commented:
Okay -- updated to switch argument order to all new functions which cleans up things quite a bit!

ppetrosh commented on ReplayMessenger.device.lib.nut:
```diff
@@ -93,6 +109,18 @@ class ReplayMessenger {
+    }
+
+    function afterReplay(loggerName, callback = null) {
+      _afterReplay[loggerName] <- callback;
```
callback = null, loggerName = "default"

mm-gmbd committed:
Bugfix

ppetrosh commented on ReplayMessenger.device.lib.nut:
```diff
@@ -93,6 +109,18 @@ class ReplayMessenger {
         return _mm.send(messageName, data, null, _retryInterval, metadata);
     }
 
+    function beforeReplay(callback = null, loggerName = "default") {
```
Thanks!! Looks much better, doesn't it?

mm-gmbd replied:
Agreed!

ppetrosh commented on ReplayMessenger.device.lib.nut:
```diff
@@ -108,38 +136,54 @@ class ReplayMessenger {
-            logger.erase(addr);
-            message.metadata.addr = null;
-            _scheduleRetryIfConnected();
+        if ("metadata" in message && "loggerName" in message.metadata) {
```
Just a minor cosmetic comment here:

I would extract this into a separate method:

```
function _isRecognized(message) {
    return "metadata" in message && "loggerName" in message.metadata;
}
```
and do a simple check here:

```
    if (! _isRecognized(message)) {
        return; // skip all non replay messenger
    }
```

ppetrosh commented on ReplayMessenger.device.lib.nut:
```diff
@@ -108,38 +136,54 @@ class ReplayMessenger {
+                    "metadata": message.metadata //add metadata to message because a user may have something important here
+                }
+
+                //TODO: Should we remove "loggerName" from metadata since it is redundant?
```
might be a good idea agree
